aten/src/THC/generic/THCTensorKthValue.cu

```diff
@@ -0,0 +1,122 @@
+  int64_t dims = THCudaLongTensor_nDimension(state, indices);
+  THArgCheck(dims <= MAX_CUTORCH_DIMS, 3, CUTORCH_DIM_WARNING);
+  int numDims = THCTensor_(nDimension)(state, input);
+  THArgCheck(numDims <= MAX_CUTORCH_DIMS, 4, CUTORCH_DIM_WARNING);
+
+  THArgCheck(dim >= 0 && dim < numDims, 6, "dim not in range");
+
+  int64_t sliceSize = THCTensor_(size)(state, input, dim);
+  THArgCheck(k > 0 && k <= sliceSize, 5, "k not in range for dimension");
+
+  // Build the output size, which is the dim being selected set to
+  // size 1
+  THLongStorage* kthValueSize = THCTensor_(newSizeOf)(state, input);
+  THLongStorage_set(kthValueSize, dim, 1);
+  THCTensor_(resize)(state, kthValue, kthValueSize, NULL);
+  THCudaLongTensor_resize(state, indices, kthValueSize, NULL);
+  THLongStorage_free(kthValueSize);
+
+  #define RUN_K(INDEX_T, DIM)                                             \
+    gatherKthValue<real, INDEX_T, DIM>                                    \
+      <<<grid, block, 0, THCState_getCurrentStream(state)>>>(             \
+        inputInfo,                                                        \
+        sliceSize,                                                        \
+        k,                                                                \
+        inputSlices,                                                      \
+        /* The actual dimension that the k-selection is running in */     \
+        /* may have changed from collapseDims() */                        \
+        inputInfo.strides[collapseInputDim],                              \
+        kthValueInfo,                                                     \
+        indicesInfo)
+
+  #define RUN_DIM(INDEX_T)                        \
+    if (allDims == 1) {                           \
+      RUN_K(INDEX_T, 1);                          \
+    } else if (allDims == 2) {                    \
+      RUN_K(INDEX_T, 2);                          \
+    } else if (allDims == 3) {                    \
+      RUN_K(INDEX_T, 3);                          \
+    } else {                                      \
+      RUN_K(INDEX_T, -1);                         \
+    }
+
+  #define RUN_T(INDEX_T)                                                  \
+    TensorInfo<real, INDEX_T> inputInfo =                                 \
+      getTensorInfo<real, THCTensor, INDEX_T>(state, input);              \
+    TensorInfo<real, INDEX_T> kthValueInfo =                              \
+      getTensorInfo<real, THCTensor, INDEX_T>(state, kthValue);           \
+    TensorInfo<int64_t, INDEX_T> indicesInfo =                            \
+      getTensorInfo<int64_t, THCudaLongTensor, INDEX_T>(state, indices);  \
+                                                                          \
+    /* We use these structures solely to find the offset to */            \
+    /* each slice we are operating on */                                  \
```

Comment:
I'm not sure what's going on here. Why are you setting the sizes of the dimension to 1 and then flattening the tensor? 

Also, shouldn't
`kthValueInfo.sizes[dim]` and  `indicesInfo.sizes[dim]` already be 1 because you resized them?